### PR TITLE
Reload the page when an app needs an update after being enabled

### LINKED ISSUE #16251: App manager stops working when an app needs to upgrade oC
### Env
8.0.3
PHP 5.5
No encryption

### Steps to reproduce

1. Enable a few apps until you can't any more

### Expected result

You should be able to enable as many apps as you want or be warned at some point that you need to log out and back in before being able to enable more.

### Actual result

After you've enabled a few apps, you may find yourself in a situation where you can't install any more app. You get an error message ("Error enabling the app") and the app manager stops working. You get a blank page when visiting the "enabled" or "disabled" tabs.

If you click on "Files", you'll be redirected to the upgrader, so it means an upgrade was required, but instead of being notified, the app manager stops working.

### Logs

```
{"reqId":"1d47b4b9bbcc55553481eb20a3114a58","app":"PHP","message":"preg_split() expects parameter 2 to be string, array given at lib\/private\/app.php#1230","level":0,"time":"2015-05-11T18:52:19+00:00","method":"POST","url":"\/index.php\/settings\/ajax\/enableapp.php"}
{"reqId":"1d47b4b9bbcc55553481eb20a3114a58","app":"PHP","message":"Invalid argument supplied for foreach() at lib\/private\/app.php#1233","level":0,"time":"2015-05-11T18:52:19+00:00","method":"POST","url":"\/index.php\/settings\/ajax\/enableapp.php"}
{"reqId":"59e504cb0234d7c803e27b74c2714aea","app":"PHP","message":"preg_split() expects parameter 2 to be string, array given at lib\/private\/app.php#1230","level":0,"time":"2015-05-11T18:52:19+00:00","method":"GET","url":"\/index.php\/settings\/ajax\/navigationdetect.php?app=files_locking"}
{"reqId":"59e504cb0234d7c803e27b74c2714aea","app":"PHP","message":"Invalid argument supplied for foreach() at lib\/private\/app.php#1233","level":0,"time":"2015-05-11T18:52:19+00:00","method":"GET","url":"\/index.php\/settings\/ajax\/navigationdetect.php?app=files_locking"}
```

### PR DESCRIPTION
**Pull request by nickvergessen**
Fix #16251

@PVince81 @oparoz 

Now displayes a notification and then reloads the page for you:

apps_-_owncloud_-_2015-10-02_12 02 32

### CODE REVIEW

**Comment by MorrisJobke** [comment]
That is really nice :)

**Comment by karlitschek** [comment]
good :+1:

**Comment by MorrisJobke** [comment]
I would do this inline - right where error messages are also shown. Let me try it.

**Comment by MorrisJobke** [comment]
I added the notification inline and added also a button to reload.

And I tested it with IE8 ;)

bildschirmfoto von 2015-10-02 14-09-25


@nickvergessen @owncloud/designers Does this look better? Or should I remove the button again? I just thought that it would be nice to see, that an action by the user is required.

**Comment by nickvergessen** [comment]
I think it should not be green, another action is required, and anything else on the page will not work anymore.
That is why I used the global thing instead of the inline.

**Comment by MorrisJobke** [comment]
> I think it should not be green

yellow/orange?

**Comment by MorrisJobke** [comment]
I think a notification is the wrong thing for this.

**Comment by MorrisJobke** [comment]
Or we fade everything out (because it doesn't work anyway) and then print "app upgrade detected - will reload in 5 seconds"

**Comment by MorrisJobke** [comment]
Something like: 

bildschirmfoto am 2015-10-02 um 15 31 46

?

**Comment by MorrisJobke** [comment]
@nickvergessen @owncloud/designers Next round :)

**Comment by MorrisJobke** [comment]
And of course tested in IE8 and IE9 ;)

**Comment by nickvergessen** [comment]
Can we reuse the modal from the remote sharing stuff?

**Comment by oparoz** [comment]
I find the inline notification the best:
* The admin can't miss it as he's just pressed the enable/update button
* There is a button for easy reloading
* It's not blocking the workflow 

Now, if an update can take down the whole instance while it waits for the reload to happen, then I think the modal window has to be chosen. I would still add a button though. Who has time to wait 5 seconds? ;)

**Comment by nickvergessen** [comment]
> Now, if an update can take down the whole instance while it waits for the reload to happen,

This is the case, any pressing of another enable/disable button will not work.

**Comment by oparoz** [comment]
> This is the case, any pressing of another enable/disable button will not work.

OK for app installs, but can't we still navigate to other parts of ownCloud? I guess it would introduce other problems where an admin forgets to reload the page and then complains that something bad is happening.

**Comment by nickvergessen** [comment]
Well any page refresh will bring up the update screen, so we are not breakinh anyything.
Just maaking sure the instance can be used again asap

**Comment by MorrisJobke** [comment]
> Can we reuse the modal from the remote sharing stuff?

Will do.

**Comment by MorrisJobke** [comment]
I updated this to the OC.dialogs:

bildschirmfoto am 2015-10-05 um 08 48 10


@nickvergessen @oparoz Is this okay now?

**Comment by nickvergessen** [comment]
Fine by me, I would use "Reload now" instead of "Ok", but that's nitpicking ;)

**Comment by MorrisJobke** [comment]
> Fine by me, I would use "Reload now" instead of "Ok", but that's nitpicking ;)

Not easy possible. Would require to redo the dialogs API :see_no_evil:

File: settings/js/apps.js [hunk 1]
```diff
@@ -390,6 +398,20 @@ OC.Settings.Apps = OC.Settings.Apps || {
 			.text('');
 	},
 
+	showReloadMessage: function(appId) {
+		OC.dialogs.info(
+			t(
+				'files_sharing',
```
Comment (DeepDiver1975) [inline settings/js/apps.js]: settings - not files_sharing

File: settings/js/apps.js [hunk 2]
```diff
@@ -390,6 +398,20 @@ OC.Settings.Apps = OC.Settings.Apps || {
...
+				'The app needs to be updated. This page will be reloaded in 5 seconds.'
+			),
+			t('files_sharing','App update'),
```
Comment (DeepDiver1975) [inline settings/js/apps.js]: settings

**Comment by MorrisJobke** [comment]
@DeepDiver1975 Fixed

**Comment by oparoz** [comment]
> Would require to redo the dialogs API

DO IT! :)

**Comment by oparoz** [comment]
- [ ] What happens if I click on the x and not on "OK"
- [ ] What happens if I press back in my browser

**Comment by MorrisJobke** [comment]
> DO IT! :)

Not two weeks before a release ;)

**Comment by MorrisJobke** [comment]
> What happens if I click on the x and not on "OK"

Then you can still do something (within the 5 seconds time window) - there is no way to handle the "close dialog" properly. And I don't want to introduce again more glue code. It should just provide a better way than the yellow notification.

> What happens if I press back in my browser

You go back. Most likely it will reload the page and then you see the upgrade page ;)

**Comment by DeepDiver1975** [comment]
> DO IT! :)

As in throw the crap away and build something useful? Contributions in the scope of 9.0 are very much welcome. :wink:

**Comment by rullzer** [comment]
Works.
Seems like the best we can get for 8.2.
:+1:

**Comment by MorrisJobke** [comment]
@nickvergessen @oparoz Can I ask you to do a final review?

**Comment by nickvergessen** [comment]
Just tested it and it works okay.

However as an admin I wonder if the app has been enabled now or not.
That is not visible in the dialog, so I'd adjust that message to something like:
`The app was enabled, but needs to be updated. You will be redirected to the update page in 5 seconds.`

other then that :+1:

**Comment by MorrisJobke** [comment]
@nickvergessen Done.

**Pull request merged**